src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.InFlightSolution.cs

```diff
@@ -0,0 +1,141 @@
+    {
+        /// <summary>
+        /// Wrapper around asynchronously produced solution.  The computation for producing the solution will be
+        /// canceled when the number of in-flight operations using it goes down to 0.
+        /// </summary>
+        public sealed class InFlightSolution
+        {
+            private readonly RemoteWorkspace _workspace;
+
+            public readonly Checksum SolutionChecksum;
+
+            private readonly CancellationTokenSource _cancellationTokenSource = new();
+
+            /// <summary>
+            /// Background work to just compute the disconnected solution associated with this <see cref="SolutionChecksum"/>
+            /// </summary>
+            private readonly Task<Solution> _disconnectedSolutionTask;
+
+            /// <summary>
+            /// Optional work to try to elevate the solution computed by <see cref="_disconnectedSolutionTask"/> to be
+            /// the primary solution of this <see cref="RemoteWorkspace"/>.
+            /// </summary>
+            private Task<Solution>? _primaryBranchTask;
+
+            /// <summary>
+            /// Initially set to 1 to represent the operation that requested and is using this solution.  This also
+            /// allows us to use 0 to represent a point that this solution computation is canceled and can not be
+            /// used again.
+            /// </summary>
+            public int InFlightCount { get; private set; } = 1;
+
+            public InFlightSolution(
+                RemoteWorkspace workspace,
+                Checksum solutionChecksum,
+                Func<CancellationToken, Task<Solution>> computeDisconnectedSolutionAsync,
+                Func<Solution, CancellationToken, Task<Solution>>? updatePrimaryBranchAsync)
+            {
+                _workspace = workspace;
+                SolutionChecksum = solutionChecksum;
+
+                _disconnectedSolutionTask = computeDisconnectedSolutionAsync(_cancellationTokenSource.Token);
+
+                // If we were asked to make this the primary workspace, then kick off that work immediately as well.
+                TryKickOffPrimaryBranchWork(updatePrimaryBranchAsync);
+            }
+
+            /// <summary>
+            /// Allow the RemoteWorkspace to try to elevate this solution to be the primary solution for itself.  This
+            /// commonly happens because when a change happens to the host, features may kick off immediately, creating
+            /// the disconnected solution, followed shortly afterwards by a request from the host to make that same
+            /// checksum be the primary solution of this workspace.
+            /// </summary>
+            /// <param name="updatePrimaryBranchAsync"></param>
+            public void TryKickOffPrimaryBranchWork(Func<Solution, CancellationToken, Task<Solution>>? updatePrimaryBranchAsync)
+            {
+                if (updatePrimaryBranchAsync is null)
+                    return;
```

Comment:
no need to actually do anything with moving this to the primary branch if the caller isn't asking for that.